File: server/dtos/work.go
package dtos

import (
	_ "github.com/mattn/go-sqlite3"
)

type GetWorkDto struct {
	Id          int      `json:"id"`
	StrId       string   `json:"str_id"`
	Dir         string   `json:"dir"`
	NameRu      string   `json:"name_ru"`
	NameEn      string   `json:"name_en"`
	Year        int      `json:"year"`
	Descr       string   `json:"descr"`
	BaseRu      string   `json:"base_ru"`
	BaseEn      string   `json:"base_en"`
	Width       int      `json:"width"`
	Height      int      `json:"height"`
	Type        int      `json:"type"`
	Images      []string `json:"images"`
	MaterialsEn []string `json:"materials_en"`
	MaterialsRu []string `json:"materials_ru"`
}

type EditWorkDto struct {
	Id           int      `json:"id"`
	StrId        string   `json:"str_id"`
	NameRu       string   `json:"name_ru"`
	NameEn       string   `json:"name_en"`
	BaseId       int      `json:"base_id"`
	Year         int      `json:"year"`
	Descr        string   `json:"descr"`
	BaseRu       string   `json:"base_ru"`
	BaseEn       string   `json:"base_en"`
	Width        int      `json:"width"`
	Height       int      `json:"height"`
	Type         int      `json:"type"`
	Images       []string `json:"images"`
	MaterialsIds []int    `json:"materials_ids"`
}

type AddWorkDto struct {
	StrId        string `json:"str_id"`
	NameRu       string `json:"name_ru"`
	NameEn       string `json:"name_en"`
	BaseId       int    `json:"base_id"`
	Year         int    `json:"year"`
	Descr        string `json:"descr"`
	Width        int    `json:"width"`
	Height       int    `json:"height"`
	Type         int    `json:"type"`
	Images       string `json:"images"`
	MaterialsIds []int  `json:"materials_ids"`
}

File: client/app/types/work.ts
// ─── WorkBase — shared fields for both Work and Sale ────────────────

export interface WorkBase {
    id: number
    str_id: string
    dir: string
    name_ru: string
    name_en: string
    year: number
    descr: string
    base_ru: string
    base_en: string
    width: number
    height: number
    type?: number
    images: string[]
    materials_ids: number[]
    materials_en?: string[]
    materials_ru?: string[]
}

// ─── Work (gallery) ─────────────────────────────────────────────────

export interface Work extends WorkBase {
    __type: 'GetWorkDto'
}

// ─── Sale (art-store) ───────────────────────────────────────────────

export interface Sale extends WorkBase {
    price: number
    __type: 'GetSaleDto'
}

// ─── Discriminated union ────────────────────────────────────────────

export type CommonWork = Work | Sale;

// ─── Create/Update DTOs ─────────────────────────────────────────────

export interface CreateWorkDto {
    str_id: string
    dir: string
    name_ru: string
    name_en: string
    year: number
    descr: string
    base_ru: string
    base_en: string
    width: number
    height: number
    type: number
    images: string[]
    materials_ids: number[]
}

export interface UpdateWorkResponse {
    id: number
    str_id: string
    dir: string
    name_ru: string
    name_en: string
    year: number
    descr: string
    base_ru: string
    base_en: string
    width: number
    height: number
    type: number
    images: string[]
    materials_ids: number[]
}

export interface UpdateWorkRequest {
    str_id?: string
    dir?: string
    name_ru?: string
    name_en?: string
    year?: number
    descr?: string
    base_ru?: string
    base_en?: string
    width?: number
    height?: number
    type?: number
    images?: string[]
    materials_ids?: number[]
}

// ─── Sale DTOs ──────────────────────────────────────────────────────

export interface CreateSaleDto {
    width: number
    height: number
    year: number
    price: number
    name_ru: string
    name_en: string
    base_id: number
    materials_ids: number[]
    descr_ru: string
    descr_en: string
    images: string[]
}

export interface UpdateSaleResponse {
    id: number
    str_id: string
    dir: string
    name_ru: string
    name_en: string
    base_id: number
    year: number
    descr_ru: string
    descr_en: string
    width: number
    height: number
    price: number
    images: string[]
    materials_ids: number[]
}

export interface UpdateSaleRequest {
    id: number
    name_ru: string
    name_en: string
    base_id: number
    year: number
    descr_ru: string
    descr_en: string
    width: number
    height: number
    price: number
    images: string[]
    materials_ids: number[]
}
